### ISSUE TITLE
Share a host path in Deis

### ISSUE THREAD

**Issue by arkkanoid** [deis/controller#858]
Hi, 
It's possible to share a host path in a container deployed with Deis?
Like: http://kubernetes.io/docs/user-guide/volumes/#hostpath

**Comment by bacongobbler** [comment]
At this time there is no way for you to configure and mount a host path into your application. Arguably I think it should stay that way as that grants any developer access to your host, and could potentially bring down your kubelet, docker daemon, kube proxy etc.

Is there a certain use case that you are trying to accomplish?

**Comment by arkkanoid** [comment]
Multiple reasons, 
Using a shared filesystem I could:
- Save data content that is generated when the application runs. I want to maintain this new data if I deploy a new release.

- For some deploys the language dependencies doesn't change. Maybe I want to look the changes in real time, without run a new application release.

About your security considerations, if I only share a particular path in the container, an user could access to the whole filesystem?

**Comment by bacongobbler** [comment]
To address your concerns:

> Save data content that is generated when the application runs. I want to maintain this new data if I deploy a new release.

When the application is rebooted, that data cannot be guaranteed to be there as your application may have been scheduled to a new host. We had that same problem in v1 where postgres used a data conatiner, but when rescheduled to a new host, data was lost. When it rebooted again and jumped on the original host, it bootstrapped itself with the old data. For that reason I strongly not suggest going this route.

> For some deploys the language dependencies doesn't change. Maybe I want to look the changes in real time, without run a new application release.

I'm not quite sure I understand what this would require host mounts. Can you elaborate on this?

> About your security considerations, if I only share a particular path in the container, an user could access to the whole filesystem?

What I mean is that if we enable users to mount host volumes into their apps, they can mount places like /var/run/docker.sock and start running containers on the host, shut them down, grab a running shell in existing containers etc. The attack surface is high.

**Comment by arkkanoid** [comment]
> To address your concerns:
> 
> Save data content that is generated when the application runs. I want to maintain this new data if I deploy a new release.
> When the application is rebooted, that data cannot be guaranteed to be there as your application may have been scheduled to a new host. We had that same problem in v1 where postgres used a data container on the host, but when rescheduled to a new host, that data container was no longer present so it started a new data container and fetched fresh from ceph. When it rebooted again and jumped on the original host, it bootstrapped itself with the old data container, losing any data since the last reboot, effectively rolling back to its original state. For that reason I strongly not suggest going this route. If you want to store data, use an external persistent storage service like S3, GCS, or minio.

But using a shared filesystem like Ceph, GlusterFS or AWS EFS we can have the same content in all hosts. Using a Hostpath volume the data would be available inside the container in any node. 
 
> For some deploys the language dependencies doesn't change. Maybe I want to look the changes in real time, without run a new application release.
> I'm not quite sure I understand why this would require host mounts. Can you elaborate on this?

If you're developing an App with Deis and you want to see the result you've to deploy each time the app. Using interpreted languages most code changes don't require re-deploy the app. If the code directory is in a hostpath maybe I can change the code without re-deploy the app. 

> About your security considerations, if I only share a particular path in the container, an user could access to the whole filesystem?
> What I mean is that if we enable users to mount host volumes into their apps, they can mount places like /var/run/docker.sock and start running containers on the host, shut them down, grab a running shell in existing containers etc. The attack surface is high and defeats the purpose of running a container isolated from the host. Might as well their app on the host itself rather than inside a container.

Absolutely agree, in that case I was thinking in one administrator that choose what directory a container has access.

In short, I was thinking the oportunities that this kind of volume can give us in Deis deployments.

**Comment by helgi** [comment]
I read over this and one of the suggestions was using hostPath to update code live without bundling it up into docker image.That'd be better served in a containerized dev setup. I can see the appeal of trying to do that with Deis but that sounds like a very heavy dependency.

#984 touches on very similar aspects as what you are talking about - Perhaps you can join up with @jsimnz to come up with a proposal on how the unique functionalities can be accommodated without creating an overwhelming amount of knobs and dials (configuration options)

Tho, on the other hand, the EFS, GlusterFS etc argument is rather strong - Could you describe the interface to this? How would I as an operator configure deis? What about all the other volume types beside hostPath? http://kubernetes.io/docs/user-guide/volumes/ and where do we want PVs to come in (http://kubernetes.io/docs/user-guide/persistent-volumes/)

**Comment by bacongobbler** [comment]
closing this, as we plan to handle EFS/GlusterFS mounts through deis/deis#231 rather than through exposing hostPath in the kubernetes manifest. As mentioned in https://github.com/deis/controller/issues/858#issuecomment-231454061, it is both a security and a scalability concern.
